### DISCUSSION TITLE
Using a locally installed dotnet tool

### DISCUSSION THREAD

**Discussion by Kemyke** [cake-build/cake#4289] [Q&A] [answered]
Hi Community,

I successfully installed a dotnet tool with the preprocessor directive:
`#tool dotnet:?package=dotnet-sonarscanner`

But when I want to use it from the build script, I get the error that the tool is not found. I try to use it like this:
`DotNetTool("sonarscanner", ...);`

Apparently, DotNetTool command can execute the sonarscanner tool when it is globally installed. I would prefer to install the tool locally. How can I execute the tool in this case with DotNetTool command? I cannot find any info about this online.

Thanks!

**Comment by nils-a** [comment, marked as answer]
I am afraid you are mixing 2 different "tools".

When you use the pre-processor directive `#tool`, the tool installed in the local cake [tools path](https://cakebuild.net/docs/running-builds/configuration/default-configuration-values#tools-path). This is independent of the "normal" `dotnet tool` usage which always installs tools in the [global tools path](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-global-tool) - even if that tool is installed locally.

I feel you have two solutions to your task:

## using the `#tool` pre-processor directive

The installed tool can be [resolved using `Context.Tools.Resolve`](https://cakebuild.net/docs/writing-builds/tools/tool-resolution). Thus, code like the following should work:

```cs
#tool dotnet:?package=dotnet-sonarscanner

var target = Argument("target", "Default");

Task("Default")
.Does(() => {
   var exe = Context.Tools.Resolve("dotnet-sonarscanner.exe");
   Information("exe: {0}", exe);
   StartProcess(exe, "-?");
});

RunTarget(target);

```

## using a locally installed tool

Assuming you are using the cake global tool as described in [the docs](https://cakebuild.net/docs/getting-started/setting-up-a-new-scripting-project), you'll probably have a tool manifest already, and you are restoring the local tools, too.
Thus, you could manually run `dotnet tool install dotnet-sonarscanner`, making sonarscanner available in your path, after the dotnet tools have been restored.

If you go that route, you'd be able to use the `DotNetTool("sonarscanner", ...);`
